orchagent/acltable.cpp

```diff
@@ -0,0 +1,436 @@
+    }
+
+    return status == SAI_STATUS_SUCCESS;
+}
+
+void AclTable::update(SubjectType type, void *cntx)
+{
+    SWSS_LOG_ENTER();
+
+    // Only interested in port change
+    if (type != SUBJECT_TYPE_PORT_CHANGE)
+    {
+        return;
+    }
+
+    PortUpdate *update = static_cast<PortUpdate *>(cntx);
+
+    Port &port = update->port;
+    if (update->add)
+    {
+        if (pendingPortSet.find(port.m_alias) != pendingPortSet.end())
+        {
+            sai_object_id_t bind_port_id;
+            if (gPortsOrch->getAclBindPortId(port.m_alias, bind_port_id))
+            {
+                link(bind_port_id);
+                bind(bind_port_id);
+
+                pendingPortSet.erase(port.m_alias);
+                portSet.emplace(port.m_alias);
+
+                SWSS_LOG_NOTICE("Bound port %s to ACL table %s",
+                        port.m_alias.c_str(), id.c_str());
+            }
+            else
+            {
+                SWSS_LOG_ERROR("Failed to get port %s bind port ID",
+                        port.m_alias.c_str());
+                return;
+            }
+        }
+    }
+    else
+    {
+        // TODO: deal with port removal scenario
+    }
+}
+
+// TODO: make bind/unbind symmetric
+bool AclTable::bind(sai_object_id_t portOid)
+{
+    SWSS_LOG_ENTER();
+
+    assert(ports.find(portOid) != ports.end());
+
+    sai_object_id_t group_member_oid;
+    if (!gPortsOrch->bindAclTable(portOid, m_oid, group_member_oid, stage))
+    {
+        return false;
+    }
+
+    ports[portOid] = group_member_oid;
+
+    return true;
+}
+
+bool AclTable::unbind(sai_object_id_t portOid)
+{
+    SWSS_LOG_ENTER();
+
+    sai_object_id_t member = ports[portOid];
+    sai_status_t status = sai_acl_api->remove_acl_table_group_member(member);
+    if (status != SAI_STATUS_SUCCESS) {
+        SWSS_LOG_ERROR("Failed to unbind table %" PRIu64 " as member %" PRIu64 " from ACL table: %d",
+                m_oid, member, status);
+        return false;
+    }
+    return true;
+}
+
+bool AclTable::bind()
+{
+    SWSS_LOG_ENTER();
+
+    for (const auto& portpair: ports)
+    {
+        sai_object_id_t portOid = portpair.first;
+        bool suc = bind(portOid);
+        if (!suc) return false;
+    }
+    return true;
+}
+
+bool AclTable::unbind()
+{
+    SWSS_LOG_ENTER();
+
+    for (const auto& portpair: ports)
+    {
+        sai_object_id_t portOid = portpair.first;
+        bool suc = unbind(portOid);
+        if (!suc) return false;
+    }
+    return true;
+}
+
+void AclTable::link(sai_object_id_t portOid)
+{
+    SWSS_LOG_ENTER();
+
+    ports.emplace(portOid, SAI_NULL_OBJECT_ID);
+}
+
+bool AclTable::add(shared_ptr<AclRule> newRule)
+{
+    SWSS_LOG_ENTER();
+
+    string rule_id = newRule->getId();
+    auto ruleIter = rules.find(rule_id);
+    if (ruleIter != rules.end())
+    {
+        // If ACL rule already exists, delete it first
+        if (ruleIter->second->remove())
```

Comment:
Can we have logging or something similar to the other methods if remove() fails?